### PR TITLE
KAFKA-5239: Producer buffer pool allocates memory inside a lock

### PR DESCRIPTION
**Pull request by smccauliff**
Move byte buffer allocation out of lock.
Add unit test for restoring count when OOM is thrown from byte buffer allocation.

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/BufferPool.java [hunk 1]
```diff
@@ -50,7 +50,7 @@
     private final ReentrantLock lock;
     private final Deque<ByteBuffer> free;
     private final Deque<Condition> waiters;
-    /** This memory is accounted for separately from the poolable buffers in free. */
+    /** Poolable + non-poolable memory. */
     private long availableMemory;
```
Comment (becketqin) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/BufferPool.java]: If we change the definition of the available memory here, we may have to change a few other places as well. e.g. `deallocate()`, `unallocatedMemory()`, etc.

Comment (smccauliff) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/BufferPool.java]: I added the original comment; it is not correct.

Comment (smccauliff) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/BufferPool.java]: Lines 234-236 confused me.  But I see that freeUp() will potentially add the count of bytes back into availableMemory.

Comment (becketqin) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/BufferPool.java]: Thanks for the explanation. It seems the comment were still not updated? Have you pushed the change?

Comment (smccauliff) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/BufferPool.java]: /** Poolable + non-poolable memory. */ is the correct comment.  No?

**Comment by becketqin** [comment]
Thanks for the patch. LGTM.

**Comment by ijuma** [comment]
Not sure if it's related, but the following test failed in Jenkins: 

```text
java.lang.AssertionError: available memory8
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.apache.kafka.clients.producer.internals.BufferPoolTest.testBlockTimeout(BufferPoolTest.java:191)
```

https://builds.apache.org/blue/organizations/jenkins/kafka-trunk-jdk8/detail/kafka-trunk-jdk8/1660/tests

**Comment by becketqin** [comment]
@ijuma I ran the tests a couple of times locally and cannot reproduce the issue. It seems an intermittent failure related to the `delayedDeallocate`. It does not look related to this patch.

**Pull request closed without merging**